plugin/completion.py

```diff
@@ -136,24 +136,35 @@ def _view_language(self, config_name: str) -> Optional[str]:
         return languages.get(config_name) if languages else None
 
     def register_trigger_chars(self, session: Session, trigger_chars: List[str]) -> None:
-        completion_triggers = self.view.settings().get('auto_complete_triggers', []) or []  # type: List[Dict[str, str]]
-        view_language = self._view_language(session.config.name)
-        if view_language:
-            for language in session.config.languages:
-                if language.id == view_language:
-                    for scope in language.scopes:
-                        # debug("registering", trigger_chars, "for", scope)
-                        scope_trigger = next(
-                            (trigger for trigger in completion_triggers if trigger.get('selector', None) == scope),
-                            None
-                        )
-                        if not scope_trigger:  # do not override user's trigger settings.
-                            completion_triggers.append({
-                                'characters': "".join(trigger_chars),
-                                'selector': scope
-                            })
-
-            self.view.settings().set('auto_complete_triggers', completion_triggers)
+        completion_triggers = self.view.settings().get('auto_complete_triggers') or []  # type: List[Dict[str, str]]
+
+        scope = main_scope_name(self.view)
+        # debug("registering", trigger_chars, "for", scope)
+
+        existing_scope_trigger = next(
+            (trigger for trigger in completion_triggers if trigger.get('selector', None) == scope),
+            None
+        )
+
+        if not existing_scope_trigger:  # do not override user's trigger settings.
+
+            def selector_for_char(char, default_scope):
+                # to avoid tiggering AC forever on commit_completion
+                # some trigger chars need to be provided a better selector
+                char_scopes = {
+                    # fixes https://github.com/sublimelsp/LSP/issues/933
+                    "\"": "punctuation.definition.string.begin"
```

Comment:
While I might agree that using selectors is the way forward, this particular implementation looks hastily written. I don't want a repeat of the previous situation where we added many workarounds for text edits. It seems like less functionality, but somehow we have more lines of code.

For a starter, I don't understand why `selector_for_char` needs to be a lambda. And it has no type hints.

Consider discussing this change in #927.

Ideally a `ServerHandler` (`LanguageHandler`?) would provide this base package with the necessary selectors.

Reply:
> I don't want a repeat of the previous situation where we added many workarounds for text edi

I agree. This code here can easily get out of control an become something like this:
https://github.com/sublimelsp/LSP/blob/24cee140fd7d02a29e82a139b918bef77d89f6fb/plugin/core/completion.py#L39-L50

As for the type hints, I haven't installed pyls-mypy correctly (I installed it, but for some reason it doesn't work, that is the reason I missed to add type hints on some places).

>Ideally a ServerHandler (LanguageHandler?) would provide this base package with the necessary selectors.

Moving this logic to a LanguageHandler/ServerHandler sounds like a good way to go. :)

I am gonna close this PR until we first discuss more about the possible solutions.